src/libpmemset/source.c

```diff
@@ -158,12 +164,18 @@ pmemset_source_from_temporaryW(struct pmemset_source **src, const wchar_t *dir)
 #endif
 
 /*
- * pmemset_source_delete -- not supported
+ * pmemset_source_delete -- delete pmemset_source structure
  */
 int
 pmemset_source_delete(struct pmemset_source **src)
 {
-	return PMEMSET_E_NOSUPP;
+	if ((*src)->type == PMEMSET_SOURCE_FILE)
```

Comment:
I suspect we are going to have many places that are source-type dependent. I'd prefer to have an array of function pointers instead of ifs all over the place.

Reply:
I do not. 

if it is so simple as this it is easier to navigate without function arrays. If we will have a lot of such code then yes - but for now I will leave it as it is.

Reply:
Temporary things tend to be permanent. e.g., I gave in with the `ifdefs` in libpmem2 and I think that was a mistake.

Even right now, it's not obvious to me that you are not forgetting to deallocate resources for other source types.

Reply:
This function could look in this way 
```
memset_source_delete(struct pmemset_source **src)
{
	return PMEMSET_E_NOSUPP;
	if ((*src)->type != PMEMSET_SOURCE_PMEM2) 
		Free((*src)->value.filepath);

	Free(*src);
	*src = NULL;

	return 0;
}
```

As all sources other then pmem2 will have a path to file or directory - I really don't see a reason to obfuscate so simple a function with function pointers

Reply:
That's even worse - the point of a tagged union is to ensure that you never access fields that were set by a different source.

The fact that this is simple right now is not an argument - by that logic, everything would be a big tangle of ifs.